clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -5637,6 +5638,27 @@ static bool EvaluateUnaryTypeTrait(Sema &Self, TypeTrait UTT,
     return false;
   case UTT_IsTriviallyEqualityComparable:
     return isTriviallyEqualityComparableType(Self, T, KeyLoc);
+  case UTT_IsImplicitLifetime: {
+    DiagnoseVLAInCXXTypeTrait(Self, TInfo,
+                              tok::kw___builtin_is_implicit_lifetime);
+    QualType UnqualT = T->getCanonicalTypeUnqualified();
+    if (UnqualT->isScalarType())
```

Comment:
Are there any extension types that should be implicit lifetime? Like `typedef int v4si __attribute__ ((vector_size (16))); static_assert(__builtin_is_implicit_lifetime(v4si));`

Reply:
Honestly, I have no idea. Given that we recently came to agreement that other stakeholders should have a say about our extensions, I wonder what libc++ thinks of this.
CC @AaronBallman @ldionne @philnik777

Reply:
We definitely have to support (some) extensions, but hopefully they shouldn't be too onerous. Things like extended scalar types (all the various floating-point types, int128_t, etc) and vector types have to be supported. But also things like `_Complex`, sizeless types, zero-sized arrays, flexible array members in structures, etc.

Reply:
I updated the patch with more tests for FP and extensions. I discovered that vector types are not considered implicit-lifetime, so I made them so. Here are tests I left out, because Clang told me ~~to go away~~ that they are not supported on the target or require special compiler flags (matrix). I took that as going over the board w.r.t. other type traits tests.
```cpp
typedef float m4x4_t __attribute__((matrix_type(4, 4)));

static_assert(__builtin_is_implicit_lifetime(__SVInt8_t));
static_assert(__builtin_is_implicit_lifetime(__rvv_int8mf8_t));
static_assert(__builtin_is_implicit_lifetime(half));
static_assert(__builtin_is_implicit_lifetime(__float128));
static_assert(__builtin_is_implicit_lifetime(__ibm128));
static_assert(__builtin_is_implicit_lifetime(__externref_t));
static_assert(__builtin_is_implicit_lifetime(__externref_t[2]));
static_assert(__builtin_is_implicit_lifetime(m4x4_t));
```
@AaronBallman Let me know if something is surprising to you, or if I forgot anything.

Reply:
> Honestly, I have no idea. Given that we recently came to agreement that other stakeholders should have a say about our extensions, I wonder what libc++ thinks of this. CC @AaronBallman @ldionne @philnik777

Me, @AaronBallman, and @philnik777 discussed that yesterday during Aaron's office hours. As I understood it, there are no concerns from libc++ side here.